include/verilated_vpi.cpp

```diff
@@ -2812,13 +2824,1086 @@ vpiHandle vpi_put_value(vpiHandle object, p_vpi_value valuep, p_vpi_time /*time_
+                        vectors[i * words + j].bval = 0x00;
+                    }
+
+                    index = ++index % size;
+                }
+            } else {
+                for (int i = 0; i < num; i++) {
+                    for (int j = 0; j < words; j++) {
+                        vectors[i * words + j].aval = ptr[index * words + j];
+                        vectors[i * words + j].bval = 0x00;
+                    }
+
+                    index = (--index + size) % size;
+                }
+            }
+        }
+
+        return;
+    } else if (arrayvalue_p->format == vpiRawFourStateVal) {
+        PLI_BYTE8 *value_ptr;
+
+        if (arrayvalue_p->flags & vpiUserAllocFlag) {
+            value_ptr = arrayvalue_p->value.rawvals;
+        } else {
+            int ngroups;
+
+            switch (varp->vltype()) {
+                case VLVT_UINT8:
+                    ngroups = 1; break;
+                case VLVT_UINT16:
+                    ngroups = 2; break;
+                case VLVT_UINT32:
+                    ngroups = 4; break;
+                case VLVT_UINT64:
+                    ngroups = 8; break;
+                case VLVT_WDATA:
+                    ngroups = VL_WORDS_I(varp->packed().elements()) * sizeof(WData); break;
+                default:
+                    ngroups = 0; break;
+            }
+
+            const int words = VL_WORDS_I(ngroups * 8 * num);
+            if (VL_UNCOVERABLE(words >= VL_VALUE_STRING_MAX_WORDS)) {
+                VL_FATAL_MT(__FILE__, __LINE__, "",
+                            "vpi_get_value_array with more than VL_VALUE_STRING_MAX_WORDS; "
+                            "increase and recompile");
+            }
+
+            value_ptr = (PLI_BYTE8*)out_ptr;
+            arrayvalue_p->value.rawvals = value_ptr;
+        }
+
+        if (varp->vltype() == VLVT_UINT8) {
+            CData *ptr = reinterpret_cast<CData*>(vop->varDatap());
+
+            struct s_rawvals {
+                uint8_t aval;
+                uint8_t bval;
+            } *rawvals = reinterpret_cast<s_rawvals*>(value_ptr);
+
+            if (vop->rangep()->left() < vop->rangep()->right()) {
+                for (int i = 0; i < num; i++) {
+                    rawvals[i].aval = ptr[index];
+                    rawvals[i].bval = 0x00;
+
+                    index = ++index % size;
+                }
+            } else {
+                for (int i = 0; i < num; i++) {
+                    rawvals[i].aval = ptr[index];
+                    rawvals[i].bval = 0x00;
+
+                    index = (--index + size) % size;
+                }
+            }
+        } else if (varp->vltype() == VLVT_UINT16) {
+            SData *ptr = reinterpret_cast<SData*>(vop->varDatap());
+
+            struct s_rawvals {
+                uint16_t aval;
+                uint16_t bval;
+            } *rawvals = reinterpret_cast<s_rawvals*>(value_ptr);
```

Comment:
Nothing should need/use this a/b uint16_t format.  Verilator doesn't internally, and IEEE AFAIK uses only it's uint32_t a/b structure.  So convert directly to output format.

Reply:
For vpiRawFourStateVal, the IEEE Std 1800-2023 on page 1107, specifies the following:

> Using the *rawvals field of the union above, interleaved according to the following structure:
> ```c
> struct {
>    PLI_BYTE8 avalbits[ngroups];
>    PLI_BYTE8 bvalbits[ngroups];
>}
> ```
> Each array element occupies ngroups*2 bytes stored consecutively as A/B byte groups as shown above.